code/datums/components/cell.dm

```diff
@@ -0,0 +1,199 @@
+	..()
+	RegisterSignal(parent, list(COMSIG_PARENT_ATTACKBY, COMSIG_ITEM_ATTACKED), PROC_REF(on_object_hit))
+	RegisterSignal(parent, COMSIG_CELL_ADD_CHARGE, PROC_REF(add_charge))
+	RegisterSignal(parent, COMSIG_CELL_USE_CHARGE, PROC_REF(use_charge))
+	RegisterSignal(parent, COMSIG_CELL_CHECK_CHARGE, PROC_REF(has_charge))
+	RegisterSignal(parent, COMSIG_CELL_START_TICK_DRAIN, PROC_REF(start_drain))
+	RegisterSignal(parent, COMSIG_CELL_STOP_TICK_DRAIN, PROC_REF(stop_drain))
+	RegisterSignal(parent, COMSIG_CELL_REMOVE_CELL, PROC_REF(remove_cell))
+	RegisterSignal(parent, COMSIG_PARENT_EXAMINE, PROC_REF(on_examine))
+	RegisterSignal(parent, COMSIG_ATOM_EMP_ACT, PROC_REF(on_emp))
+
+/datum/component/cell/process()
+	use_charge(null, charge_drain)
+
+/datum/component/cell/proc/on_emp(datum/source, severity)
+	SIGNAL_HANDLER
+
+	use_charge(null, round(max_charge / severity))
+
+/datum/component/cell/proc/start_drain(datum/source)
+	SIGNAL_HANDLER
+
+	START_PROCESSING(SSobj, src)
+
+/datum/component/cell/proc/stop_drain(datum/source)
+	SIGNAL_HANDLER
+
+	STOP_PROCESSING(SSobj, src)
+
+/datum/component/cell/proc/on_examine(datum/source, mob/examiner, list/examine_text)
+	SIGNAL_HANDLER
+
+	if(!display_charge)
+		return
+
+	if((charge_examine_range != UNLIMITED_DISTANCE) && get_dist(examiner, parent) > charge_examine_range)
+		return
+
+	examine_text += "A small gauge in the corner reads \"Power: [round(100 * charge / max_charge)]%\"."
+
+/datum/component/cell/proc/on_object_hit(datum/source, obj/item/cell/attack_obj, mob/living/attacker, params)
+	SIGNAL_HANDLER
+
+	if(!hit_charge || !istype(attack_obj))
+		return
+
+	if(!cell_insert)
+		INVOKE_ASYNC(src, PROC_REF(charge_from_cell), attack_obj, attacker)
+
+	else
+		insert_cell(attack_obj, attacker)
+
+	return COMPONENT_NO_AFTERATTACK|COMPONENT_CANCEL_ITEM_ATTACK
+
+/datum/component/cell/proc/insert_cell(obj/item/cell/power_cell, mob/living/user)
+	if(inserted_cell)
+		to_chat(user, SPAN_WARNING("There's already a power cell in [parent]!"))
+		return
+
+	if(SEND_SIGNAL(parent, COMSIG_CELL_TRY_INSERT_CELL) & COMPONENT_CANCEL_CELL_INSERT)
+		return
+
+	power_cell.drop_to_floor(user)
+	power_cell.forceMove(parent)
+	inserted_cell = power_cell
+	charge = power_cell.charge
+	max_charge = power_cell.maxcharge
+
+/datum/component/cell/proc/remove_cell(mob/living/user)
+	SIGNAL_HANDLER
+
+	user.put_in_hands(inserted_cell, TRUE)
+	to_chat(user, SPAN_NOTICE("You remove [inserted_cell] from [parent]."))
+	inserted_cell = null
+	max_charge = initial_max_charge
+	charge = 0
+
+/datum/component/cell/proc/charge_from_cell(obj/item/cell/power_cell, mob/living/user)
+	if(max_charge == UNLIMITED_CHARGE)
+		to_chat(user, SPAN_WARNING("[parent] doesn't need more power."))
+		return
+
+	while(charge < max_charge)
+		if(SEND_SIGNAL(parent, COMSIG_CELL_TRY_RECHARGING, user) & COMPONENT_CELL_NO_RECHARGE)
+			return
+
+		if(power_cell.charge <= 0)
+			to_chat(user, SPAN_WARNING("[power_cell] is completely dry."))
+			return
```

Comment:
For consistency i'd advise checking the charge after do_after (or both) to make sure it's valid at that point in time

Reply:
done